### PR TITLE
Tmp Swagger API Doc

### PR DESCRIPTION
**Pull request by ShuyuanCao**
Temporary Swagger API Doc

### CODE REVIEW

File: docs/swagger.yaml [hunk 1]
```diff
@@ -0,0 +1,978 @@
...
+        required: true
+      responses:
+        "200":
+          description: "successful operation"
+          schema:
+            $ref: "#/definitions/PredictionJobResult"
+        "400":
+          description: "invalid param"
+        
+  # /pet/{petId}:
```
Comment (TFDU) [inline docs/swagger.yaml:326]: Just curious, what are these pet related APIs? Does the "pet" word have some special meaning in ML area? Or are we letting users having virtual pets in our website?

If they are commented non-related information, please remove.

File: docs/swagger.yaml [hunk 2]
```diff
@@ -0,0 +1,978 @@
...
+- "https"
+- "http"
+paths:
+  /automl/login:
+    post:
+      tags:
+      - "Authentication"
+      summary: "Login to AutoML"
+      description: ""
+      operationId: "login"
```
Comment (TFDU) [inline docs/swagger.yaml:41]: I think we need register as well.

**Review by TFDU — CHANGES_REQUESTED** [review]
Generally looks good to me.

I do have some suggestions.

First, I think we are missing register.

Second, I think we are missing Authorization response status. For most of the APIs, we should verify the user. And if the user does not have enough authorization to perform the action, we should return another response, eg.. "No Authorizaiton". Without this check UserB might have a chance to modify userA's dataset through API even if we do not expose an entry in frontend.

Third, I suggest we split up this API design, either by feature, or by user journey. Currently we are having all these core APIs together, which tends to miss some important stuff. For example, if we split up these APIs by features, it should be somewhat like below:
**[Authentication]**
  - Login - 
  - Logout - 
  - Register - 
  - Change password - 
  - Remove account -
**[User information management]** 
  - Update user inforamtionn -
**[Dataset mangement]**
...
**[ML status mangement]**
...

To split up the APIs, we have more clear structure on the APIs, and we can make the pull request as small as we can. As a result, during the iteration, we do not need to review all the APIs all the time, we only need to focus on the updated one.

Finally, I have no experience working on API design separately(In my daily work, we degisn our own API for different feature, never doing this all at once at first) So these are just my suggestions, it is totally fine if you do not take them.

**Comment by free-y** [comment]
try to reopen it

File: docs/swagger.yaml [hunk 3]
```diff
@@ -0,0 +1,978 @@
...
+      code:
+        type: "integer"
+        format: "int32"
+      type:
+        type: "string"
+      message:
+        type: "string"
+externalDocs:
+  description: "Find out more about Swagger"
+  url: "http://swagger.io"
```
Comment (free-y) [inline docs/swagger.yaml:978]: test if I can add new comment

**Comment by free-y** [comment]
> Finally, I have no experience working on API design separately(In my daily work, we degisn our own API for different feature, never doing this all at once at first) So these are just my suggestions, it is totally fine if you do not take them.
Great point! We can use ShuyuanCao's doc as example and reference, the APIs definitely need to be better organized just as you suggested.  Which way is better, split up the APIs into different files or being better organized in a single file, how do you think?

File: docs/swagger.yaml (hunk 2 above)
Comment (free-y) [reply docs/swagger.yaml:41]: Absolutely agree! And we need to think carefully about the multi-tenancy case.

File: docs/swagger.yaml (hunk 1 above)
Comment (free-y) [reply docs/swagger.yaml:326]: > Just curious, what are these pet related APIs? Does the "pet" word have some special meaning in ML area? Or are we letting users having virtual pets in our website?
> 
> If they are commented non-related information, please remove.

ShuyuanCao told me that the commented block is the original template, we can remove it once we don't need to use it as a reference any more.

**Comment by free-y** [comment]
By the way, the APIs in this doc only cover four menu items in our UI prototype, we need to finish all the API definition according to our UI prototype in the near future.
